Slash Commands

amberstarlight opened the pull request:
This PR changes all the commands over to [slash commands](https://support.discord.com/hc/en-us/articles/1500000368501-Slash-Commands-FAQ).

Given that Discord will be [putting message content behind a privileged intent](https://support-dev.discord.com/hc/en-us/articles/4404772028055-Message-Content-Privileged-Intent-for-Verified-Bots) this change is required. Their blog post suggests this will not affect bots that are in less than 75 servers (that applies to us) but I would like to get ahead of the curve and preempt Discord removing this exemption. Plus, slash commands are much better UI/UX!

As the old command system is gone I have removed the sole test.

I have also updated the event handling as we are now getting `interactionCreate` events.

amberstarlight committed:
Remove tests for updated command system

amberstarlight committed:
Update event handling for new commands

amberstarlight committed:
Lint utils

amberstarlight committed:
Update event files

amberstarlight committed:
Update cat command

amberstarlight committed:
Update flip command

amberstarlight committed:
Update ping command

amberstarlight committed:
Update prune command

amberstarlight committed:
Fix event handling

amberstarlight committed:
Update kick command

amberstarlight committed:
Update ban command

amberstarlight committed:
Update embed content for kick/prune commands

amberstarlight committed:
Remove jest dependency

amberstarlight commented on deploy-commands.js:
```diff
@@ -0,0 +1,36 @@
+		console.log('Started refreshing application (/) commands.');
+
+		await rest.put(
+			Routes.applicationGuildCommands(clientId, guildId),
```
To deploy slash commands we need to run `deploy-commands.js`. Currently we are deploying to the test server (line 27) but on deployment we will need to run this with `Routes.applicationCommands` to push the slash commands globally (for all guilds).

I'm not sure what the best idea is to make sure that we do this for each new command; i.e. in development we should push to to the test server, but in prod we should be deploying globally (but only need to do so if command files have changed).

Any thoughts?

amberstarlight replied:
~~As discussed we can run this on startup, checking `NODE_ENV`.~~

amberstarlight replied:
I've moved this to the `yarn start` command, instead.

amberstarlight committed:
Update utils to remove prefix from database & validation

amberstarlight committed:
Move interactionCreate to event file

amberstarlight committed:
Add thread support

amberstarlight committed:
Update reminder command

amberstarlight committed:
Remove old reminder commands

amberstarlight committed:
Update roles command

amberstarlight committed:
Update settings command and util

amberstarlight committed:
Update vote command

amberstarlight committed:
Remove legacy commands

mini-ninja-64 commented on commands/ban.js:
```diff
@@ -1,49 +1,73 @@
-	usage: "[@user]",
-	cooldown: 0,
-	guildOnly: true,
-	modOnly: true,
```
I personally think that this config driven approach is quite nice, unless there's a big reason not too I would keep it and have slash command definitions built based upon it. Would help reduce boiler plate on each command

amberstarlight replied:
We discussed this and the abstraction isn't worth it.

mini-ninja-64 commented on commands/cat.js:
```diff
@@ -1,32 +1,39 @@
 
-		getRequest(`https://cataas.com/cat${tags}`)
-			.then((data) => {
+		await getRequest(`https://cataas.com/cat${tags}`)
```
Ideally should not be mixing `await` and `.then` `.catch` syntax

mini-ninja-64 commented on commands/kick.js:
```diff
@@ -1,48 +1,55 @@
+				.setDescription('The reason for kicking this user.')
+		),
+	async execute(interaction) {
+		const member = interaction.options.getMember('target');
```
Would it be worth having a helper method for fetching these 3 based on interaction as we use it regularly

amberstarlight replied:
As discussed this is different in each function so probably not worth abstracting away.

mini-ninja-64 commented on commands/prune.js:
```diff
@@ -1,46 +1,79 @@
-				}).then(msg => {
-					setTimeout(() => msg.delete(), 5000);
+		if (amount > 1) {
+			interaction.channel.bulkDelete(amount, true)
```
Same thoughts wrt mixing async and .then / .catch syntax in the same function

mini-ninja-64 commented on commands/reminder.js:
```diff
@@ -0,0 +1,115 @@
+			ephemeral: true
+		});
+
+		if (interaction.options.getSubcommand() === 'create') { // create reminder
```
Could be cleaner to have a function for each sub command and a switch case Or look up table to prevent one mega function

Just a thought not saying you have to of course

mini-ninja-64 commented on commands/roles.js:
```diff
@@ -1,55 +1,150 @@
 		let roleList = [];
 		assignableRoles.forEach((role) => {
-			roleList.push(`**${role.name}** \u2013 ${role.id}`);
+			roleList.push(role);
```
I think `roleList` might just be a clone of `assignableRoles` if we don't modify `roleList` we can just use `assignableRoles` directly

mini-ninja-64 commented on commands/roles.js:
```diff
@@ -1,55 +1,150 @@
-			if (role.name === "@everyone") return false;
+		let userRoles = Array.from(interaction.member.roles.cache.values());
+		let cleanedUserRoles = userRoles.filter(chosenRole => {
+			if (chosenRole.name === '@everyone') return false;
```
Second time we used this call back so could store in a variable to save repeating ourselves :3

mini-ninja-64 reviewed (changes requested):
Review part 1 
Looking good!! Excited to get it merged (≧▽≦)

amberstarlight committed:
Add editorconfig

amberstarlight committed:
Update files from PR review

mini-ninja-64 commented on commands/settings.js:
```diff
@@ -1,91 +1,154 @@
-	if (item) return (item.name);
-	return "ERROR";
+	if (item) return item.name;
+	return 'ERROR';
```
should this throw an errior instead of returning a string :?

mini-ninja-64 commented on commands/settings.js:
```diff
@@ -1,91 +1,154 @@
+		})
+		.join('\n');
 
 const transformations = {
```
maybe name like settingsTransformer would be better

mini-ninja-64 commented on commands/settings.js:
```diff
@@ -1,91 +1,154 @@
+	welcome_channel_id: 'Welcome Message Channel',
 };
 
+const optionChoices = [
```
cant we just use `settingsPretifier` for this data? or maybe generate settigns pretifier using this set of data

mini-ninja-64 commented on commands/settings.js:
```diff
@@ -1,91 +1,154 @@
+			return interaction.editReply({ embeds: [settingsEmbed] });
+		}
+
+		// getSubcommand === 'update'
```
think it would be good to not fall through to update, it could be abused as someone sending illegal option would cause a settings update

switch case and function breakout like elsewhere could be nice

mini-ninja-64 commented on events/guildDelete.js:
```diff
@@ -1,18 +1,30 @@
+				)
+			)
+			.catch(log.error);
+	},
```
this could be a function as used in 2 places in the project

mini-ninja-64 reviewed (changes requested):
Review part 2 :3

amberstarlight committed:
Update guildCreate, guildDelete

amberstarlight committed:
Update settings.js, cat.js
- Rename transformations to settingsTransformer
- Throw error in idResolver instead of returning a string
- Use try...catch blocks to catch errors
- Rename settingsPrettifier to settingStringify
- Refactor settings.js getSubcommand

amberstarlight committed:
Fix error handling in cat.js

amberstarlight committed:
Fix error handling in webRequest.js

amberstarlight committed:
Update scripts to deploy slash commands

mini-ninja-64 commented on commands/prune.js:
```diff
@@ -1,46 +1,113 @@
+					],
+				});
+			}
+		}
```
TGhis could be simplified into 1 block I think like
```js
try {
  if(amount > 1) {
    // do bulky delete
  } else {
    // do single delete
  }
} catch {
}
```

mini-ninja-64 commented on commands/roles.js:
```diff
@@ -1,55 +1,182 @@
-	}
+		if (interaction.options.getSubcommand() === 'view') {
+			return interaction.editReply({ embeds: [rolesEmbed] });
+		}
```
the view block may fit nicely in switch case

mini-ninja-64 commented on commands/settings.js:
```diff
@@ -1,91 +1,169 @@
+
+		switch (interaction.options.getSubcommand()) {
+			case 'view':
+				interaction.editReply({ embeds: [settingsEmbed] });
```
I would have view as its own function too for consistency where the embed is built otherwise we end up polluting the main function building embed when we dont need to

mini-ninja-64 commented on deploy-commands.js:
```diff
@@ -0,0 +1,41 @@
+		console.log('Started refreshing application (/) commands.');
+
+		if (process.env.NODE_ENV == 'production') {
+			await rest.put(Routes.applicationCommands(clientId), {
```
Is it worth logging whether its pushing commands to global or guild?

mini-ninja-64 commented on utils/database.js:
```diff
@@ -1,22 +1,26 @@
-const fs = require("fs");
+// otherwise deploy-commands.js throws an error
+const loggerInit = require('./logger.js');
+loggerInit();
```
if logger is required for database setup in deploy-command.js this should be initialized in deploy-command.js instead of in database zone.

I would remove loggerInit(); from here and add it to deploy-command.js

mini-ninja-64 commented on utils/settings.js:
```diff
@@ -1,94 +1,74 @@
 	return roles;
 };
 
 // settings validators
 const validSettings = {
```
These are no longer settingValidators because they transform and return the new setting value for the database, instead of just validating

mini-ninja-64 reviewed (changes requested):
Almost there :3 looking really good

amberstarlight committed:
Refactor switch statements in commands

amberstarlight committed:
Update utils
- Move logger call to deploy-commands
- Rename validSettings object

amberstarlight committed:
Refactor scope of multi-option commands

amberstarlight committed:
Fix fall-through in roles.js

mini-ninja-64 reviewed (approved):
<img src="https://media2.giphy.com/media/cW64pEEZe0YZa/giphy.gif"/>
